Wikipedia/View Controllers/ShareCard/WMFShareOptionsController.m

```diff
@@ -0,0 +1,346 @@
+    
+    [shareOptionsView mas_remakeConstraints:^(MASConstraintMaker *make) {
+        make.width.equalTo(containingView.mas_width);
+        make.centerX.equalTo(containingView.mas_centerX);
+
+        make.top.equalTo(containingView.mas_bottom);
+    }];
+    
+    [shareOptionsView layoutIfNeeded];
+
+    [shareOptionsView mas_remakeConstraints:^(MASConstraintMaker *make) {
+        make.width.equalTo(containingView.mas_width);
+        make.centerX.equalTo(containingView.mas_centerX);
+    
+        make.bottom.equalTo(self.containerViewController.mas_bottomLayoutGuide);
+    }];
+    
+    [UIView animateWithDuration:0.40 delay:0.0 usingSpringWithDamping:0.8 initialSpringVelocity:0.0 options:0 animations:^{
+        
+        [shareOptionsView layoutIfNeeded];
+        self.grayOverlay.alpha = 1.0;
+        
+    } completion:^(BOOL finished) {
+        UITapGestureRecognizer* tapForCardOnCardImageViewRecognizer = [[UITapGestureRecognizer alloc]
+                                                                       initWithTarget:self action:@selector(respondToTapForCardGesture:)];
+        UITapGestureRecognizer* tapForCardOnButtonRecognizer = [[UITapGestureRecognizer alloc]
+                                                                initWithTarget:self action:@selector(respondToTapForCardGesture:)];
+        UITapGestureRecognizer* tapForTextRecognizer = [[UITapGestureRecognizer alloc]
+                                                        initWithTarget:self action:@selector(respondToTapForTextGesture:)];
+        [self.shareOptions.cardImageViewContainer addGestureRecognizer:tapForCardOnCardImageViewRecognizer];
+        [self.shareOptions.shareAsCardLabel addGestureRecognizer:tapForCardOnButtonRecognizer];
+        [self.shareOptions.shareAsTextLabel addGestureRecognizer:tapForTextRecognizer];
+    }];
+}
+
+- (void)dismissShareOptions {
+    
+    UIView* containingView = self.containerViewController.view;
+
+    [self.shareOptions mas_remakeConstraints:^(MASConstraintMaker *make) {
+        make.width.equalTo(containingView.mas_width);
+        make.centerX.equalTo(containingView.mas_centerX);
+        
+        make.top.equalTo(containingView.mas_bottom);
+    }];
+
+    [UIView animateWithDuration:0.40 delay:0.0 usingSpringWithDamping:0.8 initialSpringVelocity:0.0 options:0 animations:^{
+        self.grayOverlay.alpha = 0.0;
+        [self.shareOptions layoutIfNeeded];
+    } completion:^(BOOL finished) {
+        [self.grayOverlay removeFromSuperview];
+        [self.shareOptions removeFromSuperview];
+        self.grayOverlay = nil;
+        self.shareOptions = nil;
+    }];
+}
+
+
+#pragma mark - Tap Gestures
+
+- (void)respondToDimAreaTapGesture:(UITapGestureRecognizer*)recognizer {
+    [self.shareFunnel logAbandonedAfterSeeingShareAFact];
+    [[WMFImageController sharedInstance] cancelFetchForURL:[NSURL wmf_optionalURLWithString:self.article.imageURL]];
```

Comment:
shouldn't we put this (and any other tear-down logic) into the `dismissShareOptions` method?